docs/2-building-blocks-of-oop-part-1/hometask/src/Shape.ts

```diff
@@ -1,3 +1,32 @@
+import { Point } from './Point';
 export abstract class Shape {
+  private color: string;
+  private filled: boolean;
+  points: Point[];
+  constructor(points: Point[]);
+  constructor(points: Point[], color: string, filled: boolean);
+  constructor(points: Point[], color?: string, filled?: boolean) {
+    if (points.length < 3) {
+      throw new Error('shape has least 3 points');
+    }
+    this.points = points;
+    this.color = color || 'green';
+    this.filled = typeof filled === 'boolean' ? filled : true;
+  }
+  toString() {
+    const pointsStrArr: string[] = [];
+    for (let point of this.points) {
+      pointsStrArr.push(point.toString());
+    }
```

Comment:
Hi @siyuejiwang , this part seems not necessary as when you call `Array.prototype.join`, for each element of `Array`, `toString` will be invoked first, meaning that you don't have to use `toString` manually and `points.join(', ')` can do the same thing.

Reply:
ok, thank you, got it